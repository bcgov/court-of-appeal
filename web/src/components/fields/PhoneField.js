import React, { Component } from 'react';
import {INVALID_PHONE_MSG} from "../../helpers/constants";

class PhoneField extends Component {

    constructor(props) {
        super(props);
        this.handleFieldChange= this.handleFieldChange.bind(this);
        this.validate = this.validate.bind(this);
        this.isValidPhoneNumber = this.isValidPhoneNumber.bind(this);
        this.clearClass = this.clearClass.bind(this);
        this.state = {
            className: this.props.value ? "form-field preloaded" : "form-field",
            isValid: false,
            title: '',
            errorMessage: ''
        };
    }

    render() {
        let autofocus = this.props.autofocus ? true : false;
        return (
                <div>
                    <input
                        id={this.props.id}
                        name={this.props.name}
                        autoFocus={autofocus}
                        ref={this.props.ref}
                        value={this.props.value || ''}
                        onChange={this.handleFieldChange}
                        maxLength={17}
                        size={30}
                        placeholder={this.props.placeHolder}
                        readOnly={this.props.readOnly}
                        onBlur={this.validate}
                        className={this.state.className}
                        onFocus={this.clearClass}
                    />
                    <div className="error-message"> {this.state.errorMessage}</div>
                </div>


        );
    }

    componentDidMount() {
        if (this.props.value) {
            let isValid = this.validate();
            if (isValid && this.props.value) {
                this.setState({className: "form-field preloaded"});
            }
        }
    }

    /**
     * Eliminate all but valid Canadian phone characters are used eg +1(250)385-9999
     * @param e
     */
    handleFieldChange(e) {
        e.target.value = e.target.value.replace(/[^+\d-)( ]/g,'');
        this.props.handleFieldChange(e);
    }

    validate() {
        if (!this.props.value || this.isValidPhoneNumber())
        {
            this.setState({className: "form-field ", title: "", errorMessage: ""}, this.props.validate(true, 'phone'));
            return true;
        } else {
            this.setState({className: "form-field invalid-field", title: INVALID_PHONE_MSG, errorMessage: INVALID_PHONE_MSG}, this.props.validate(false, 'phone'));
            return false;
        }
    }

    isValidPhoneNumber() {
        return this.props.value.match(/^(\+)?(1)?[ ]?\(\d{3}\)[ ]?\d{3}[- ]?\d{4}$/) != null ||
        this.props.value.match(/^(\+)?(1)?[- ]?\d{3}[- ]?\d{3}[- ]?\d{4}$/) != null ||
        this.props.value.match(/^(\+)?(1)?\d{3}-\d{3}-\d{4}$/) != null;
    }

    clearClass() {
        this.setState({className: "form-field", title: "", errorMessage: ""});
    }
}
export default PhoneField;
