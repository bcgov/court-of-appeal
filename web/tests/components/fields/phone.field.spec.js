import React from 'react';
import PhoneField from '../../../src/components/fields/PhoneField';
import renderer from 'react-test-renderer';

test('should only allow numeric, dash, and parentheses entries', ()=> {

    let value = '';
    let handleFieldChange = (e) => { value = e.target.value; };

    const phoneField = renderer.create(
        <PhoneField id="number-field" value={value}
                      handleFieldChange={handleFieldChange.bind(this)}/>,
    );
    let tree = phoneField.toJSON();
    expect(tree).toMatchSnapshot();

    let e = { target: { value: "42"}};
    let p = phoneField.getInstance();

    p.handleFieldChange(e);
    expect(value).toMatch('42');

    e = { target: { value: "bah" } };
    p.handleFieldChange(e);
    expect(value).toMatch('');

    e = { target: { value: "(250)8674309" } };
    p.handleFieldChange(e);
    expect(value).toMatch('(250)8674309');

    e = { target: { value: "+bah1(867)-5309)pooh!" } };
    p.handleFieldChange(e);
    expect(value).toMatch('+1(867)-5309)');

});
