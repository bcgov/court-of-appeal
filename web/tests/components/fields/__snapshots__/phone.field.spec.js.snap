// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should only allow numeric, dash, and parentheses entries 1`] = `
<div>
  <input
    autoFocus={false}
    className="form-field"
    id="number-field"
    maxLength={17}
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    size={30}
    value=""
  />
  <div
    className="error-message"
  >
     
    
  </div>
</div>
`;
